--- camlzip-1.04/Makefile	2009-10-20 15:59:55.000000000 +0200
+++ camlzip-1.04/Makefile.new	2009-10-20 16:00:31.000000000 +0200
@@ -4,10 +4,10 @@
 ZLIB_LIB=-lz
 
 # The directory containing the Zlib library (libz.a or libz.so)
-ZLIB_LIBDIR=/usr/local/lib
+ZLIB_LIBDIR=/opt/local/lib
 
 # The directory containing the Zlib header file (zlib.h)
-ZLIB_INCLUDE=/usr/local/include
+ZLIB_INCLUDE=/opt/local/include
 
 # Where to install the library.  By default: sub-directory 'zip' of
 # OCaml's standard library directory.
--- /dev/null	2009-10-20 16:35:40.000000000 +0200
+++ camlzip-1.04/META	2009-10-20 16:37:31.000000000 +0200
@@ -0,0 +1,6 @@
+name = "camlzip"
+version = "1.04"
+description = "compression library"
+archive(byte) = "zip.cma"
+archive(native) = "zip.cmxa"
+directory = "+zip"
--- camlzip-1.04/Makefile.orig	2011-07-04 18:09:00.000000000 +0200
+++ camlzip-1.04/Makefile	2011-07-04 18:10:09.000000000 +0200
@@ -56,7 +56,8 @@
 
 install:
 	mkdir -p $(INSTALLDIR)
-	cp zip.cma zip.cmi gzip.cmi zip.mli gzip.mli libcamlzip.a $(INSTALLDIR)
+	cp zip.cma zip.cmi gzip.cmi zlib.cmi zip.mli gzip.mli zlib.mli \
+	   libcamlzip.a $(INSTALLDIR)
 	if test -f dllcamlzip.so; then \
 	  cp dllcamlzip.so $(INSTALLDIR); \
           ldconf=`$(OCAMLC) -where`/ld.conf; \
@@ -66,7 +67,7 @@
         fi
 
 installopt:
-	cp zip.cmxa zip.a zip.cmx gzip.cmx $(INSTALLDIR)
+	cp zip.cmxa zip.a zip.cmx gzip.cmx zlib.cmx $(INSTALLDIR)
 
 depend:
 	gcc -MM -I$(ZLIB_INCLUDE) *.c > .depend
