--- coq-8.4pl1/kernel/univ.ml.orig	2013-11-27 15:53:01.000000000 +0100
+++ coq-8.4pl1/kernel/univ.ml	2013-11-27 15:53:20.000000000 +0100
@@ -226,7 +226,7 @@
 
 
 (* between : UniverseLevel.t -> canonical_arc -> canonical_arc list *)
-(* between u v = {w|u<=w<=v, w canonical}          *)
+(* between u v = {w |u<=w<=v, w canonical}          *)
 (* between is the most costly operation *)
 
 let between g arcu arcv =
