diff -N -r -u sks-1.1.3.orig/Makefile.local sks-1.1.3/Makefile.local
--- sks-1.1.3.orig/Makefile.local	1970-01-01 01:00:00.000000000 +0100
+++ sks-1.1.3/Makefile.local	2010-05-17 14:49:16.000000000 +0200
@@ -0,0 +1,5 @@
+LIBDB=-ldb
+MANDIR=${PREFIX}/share/man
+export PREFIX
+export LIBDB
+export MANDIR
--- sks-1.1.3.orig/Makefile	2012-04-11 04:03:25.000000000 +0200
+++ sks-1.1.3/Makefile	2013-05-30 14:40:03.000000000 +0200
@@ -47,7 +47,7 @@
 
 CAMLP4=-pp $(CAMLP4O)
 CAMLINCLUDE= -I lib -I bdb
-COMMONCAMLFLAGS=$(CAMLINCLUDE) $(OCAMLLIB) -ccopt -Lbdb -dtypes -ccopt -pthread -ccopt -pg -warn-error A
+COMMONCAMLFLAGS=$(CAMLINCLUDE) $(OCAMLLIB) -ccopt -Lbdb -dtypes -ccopt -pthread -ccopt -pg -warn-error a
 OCAMLDEP=ocamldep $(CAMLP4) 
 CAMLLIBS=unix.cma str.cma bdb.cma nums.cma bigarray.cma cryptokit.cma
 OCAMLFLAGS=$(COMMONCAMLFLAGS) -g $(CAMLLIBS)
