--- vsyml-2010-04-06.orig/makefile	2010-04-06 19:28:25.000000000 +0200
+++ vsyml-2010-04-06/makefile	2010-08-23 15:16:22.000000000 +0200
@@ -525,13 +525,13 @@
 
 # dependencies for the symbolic simulator main file on cmo cma cmx and cmxa
 $(VSYML_CMO_LST): $(VSYML_MAIN)
-	echo -n "VSYML_CMO=" > $@
-	for i in `grep -o -e '[a-zA-Z0-9_]*\.cmo' $<` ; do echo -n $$i " " >> $@ ; done
+	echo "VSYML_CMO=" | tr -d '\012' > $@
+	for i in `grep -o -e '[a-zA-Z0-9_]*\.cmo' $<` ; do echo $$i " " | tr -d '\012' >> $@ ; done
 	echo $(patsubst $(SRC_PATH)$(PATH_SEPARATOR)%.ml,%.cmo,$<) >> $@
 
 $(VSYML_CMA_LST): $(VSYML_MAIN)
-	echo -n "VSYML_CMA=" > $@
-	for i in `grep -o -e '[a-zA-Z0-9_]*\.cma' $<` ; do echo -n $$i " " >> $@ ; done
+	echo "VSYML_CMA=" | tr -d '\012' > $@
+	for i in `grep -o -e '[a-zA-Z0-9_]*\.cma' $<` ; do echo $$i " " | tr -d '\012' >> $@ ; done
 
 $(VSYML_BYTE_CMO_LST): $(VSYML_CMO_LST)
 	sed -e 's@\([a-zA-Z0-9_]*\)\.cmo@$(BYTE_PATH)$(PATH_SEPARATOR)\1.cmo@g' -e 's/VSYML_CMO/VSYML_BYTE_CMO/' $< > $@
