--- xml-light/Makefile	2003-10-12 11:16:12.000000000 +0200
+++ xml-light-2.2/Makefile	2010-01-23 20:57:57.000000000 +0100
@@ -2,7 +2,7 @@
 # http://tech.motion-twin.com
 .SUFFIXES : .ml .mli .cmo .cmx .cmi .mll .mly
 
-INSTALLDIR=`ocamlc -where`
+INSTALLDIR=`ocamlc -where`/xml-light
 CFLAGS=
 LFLAGS= -a
 LIBS=
@@ -12,6 +12,7 @@
 opt: xml-light.cmxa test_opt.exe
 
 install: all opt
+	mkdir -p "${INSTALLDIR}"
 	cp xml-light.cmxa xml-light.a xml-light.cma xml.mli xmlParser.mli dtd.mli xml.cmi xmlParser.cmi dtd.cmi xml.cmx dtd.cmx xmlParser.cmx $(INSTALLDIR)
 
 doc:
