--- coq-8.3pl4.orig/configure	2011-12-19 22:57:30.000000000 +0100
+++ coq-8.3pl4/configure	2012-03-16 11:44:55.000000000 +0100
@@ -444,7 +444,7 @@
 
 if [ "$coq_debug_flag" = "-g" ]; then
     case $CAMLTAG in
-        OCAML31*)
+        OCAML31*|OCAML4*)
             # Compilation debug flag
             coq_debug_flag_opt="-g"
             ;;
@@ -494,7 +494,7 @@
     camlp4oexec=`echo $camlp4oexec | sed -e 's/4/5/'`
 else
     case $CAMLTAG in
-        OCAML31*)
+        OCAML31*|OCAML4*)
             if [ -x "${CAMLLIB}/camlp5" ]; then
                 CAMLP4LIB=+camlp5
             elif [ -x "${CAMLLIB}/site-lib/camlp5" ]; then
@@ -538,7 +538,7 @@
       CAMLOPTVERSION=`"$nativecamlc" -v | sed -n -e 's|.*version* *\(.*\)$|\1|p' `
       if [ "`uname -s`" = "Darwin" -a "$ARCH" = "i386" ]; then
 	  case $CAMLOPTVERSION in
-	      3.09.3|3.1?*) ;;
+	      3.09.3|3.1?*|4.*) ;;
 	      *) echo "Native compilation on MacOS X Pentium requires Objective-Caml >= 3.09.3,"
 		 best_compiler=byte
 		 echo "only the bytecode version of Coq will be available."
--- coq-8.3pl4/scripts/coqmktop.ml.orig	2012-05-26 21:32:12.000000000 +0200
+++ coq-8.3pl4/scripts/coqmktop.ml	2012-05-26 21:36:35.000000000 +0200
@@ -63,6 +63,7 @@
       (src_dirs ())
       (["-I"; "\"" ^ camlp4lib ^ "\""] @
 	 ["-I"; "\"" ^ coqlib ^ "\""] @
+         ["-I"; "+compiler-libs"] @
 	 (if !coqide then ["-thread"; "-I"; "+lablgtk2"] else []))
 
 (* Transform bytecode object file names in native object file names *)
@@ -274,7 +275,7 @@
         ocamloptexec^" -linkall"
     end else
       (* bytecode (we shunt ocamlmktop script which fails on win32) *)
-      let ocamlmktoplib = " toplevellib.cma" in
+      let ocamlmktoplib = " ocamlcommon.cma ocamlbytecomp.cma ocamltoplevel.cma" in
       let ocamlcexec = Filename.concat camlbin "ocamlc" in
       let ocamlccustom = Printf.sprintf "%s %s -linkall "
         ocamlcexec Coq_config.coqrunbyteflags in
--- coq-8.3pl4/configure.orig	2012-07-18 11:31:08.353180800 +0200
+++ coq-8.3pl4/configure	2012-07-18 11:31:10.346046400 +0200
@@ -272,7 +272,7 @@
     no) 
     # First we test if we are running a Cygwin system
     if [ `uname -s | cut -c -6` = "CYGWIN" ] ; then
-	ARCH="win32"
+	ARCH=`uname -s`
     else
 	# If not, we determine the architecture
 	if test -x /bin/arch ; then
