--- lwt-2.4.0.orig/src/unix/lwt_unix.ml	2012-07-19 13:35:56.000000000 +0200
+++ lwt-2.4.0/src/unix/lwt_unix.ml	2013-08-13 15:46:12.000000000 +0200
@@ -596,6 +596,9 @@
 #if ocaml_version >= (3, 13)
   | O_SHARE_DELETE
 #endif
+#if ocaml_version >= (4, 01)
+  | O_CLOEXEC
+#endif
 
 #if windows
 
--- lwt-2.4.0.orig/src/unix/lwt_unix.mli	2012-07-19 13:35:56.000000000 +0200
+++ lwt-2.4.0/src/unix/lwt_unix.mli	2013-08-13 15:46:18.000000000 +0200
@@ -315,6 +315,9 @@
 #if ocaml_version >= (3, 13)
   | O_SHARE_DELETE
 #endif
+#if ocaml_version >= (4, 01)
+  | O_CLOEXEC
+#endif
 
 val openfile : string -> open_flag list -> file_perm -> file_descr Lwt.t
   (** Wrapper for [Unix.openfile]. *)
