--- mldonkey-3.1.2.orig/config/configure	2011-08-08 07:11:57.000000000 +0200
+++ mldonkey-3.1.2/config/configure	2012-03-13 12:52:40.000000000 +0100
@@ -4870,7 +4870,7 @@
 else
   OCAMLVERSION=`$OCAMLC -v | sed -n -e 's|.*version* *\(.*\)$|\1|p' `
   case "$OCAMLVERSION" in
-    "$REQUIRED_OCAML"*|3.12.*|3.11.*|3.10.1*|3.10.2*) ;;
+    "$REQUIRED_OCAML"*|4.*|3.1[23].*|3.11.*|3.10.1*|3.10.2*) ;;
     *)
         echo "Need build"
         BUILD_OCAML=yes
@@ -5402,7 +5402,7 @@
 
 # force usage of supported Ocaml versions
 case "$OCAMLVERSION" in
-    3.10.1*|3.10.2*|3.1*) ;;
+    3.10.1*|3.10.2*|3.1*|4.*) ;;
     *)
       if  test "$REQUIRED_OCAML" != "SVN" ; then
         echo "********  Version $REQUIRED_OCAML of Objective-Caml is required  *********" 1>&2;
--- mldonkey-3.1.2.orig/Makefile	2012-05-16 11:56:34.000000000 +0200
+++ mldonkey-3.1.2/Makefile	2012-05-25 19:24:15.000000000 +0200
@@ -5447,7 +5449,7 @@
 	$(OCAMLC) $(DEVFLAGS) $(INCLUDES) -c $<
 
 .mlcpp.ml:
-	@$(CPP) -x c -P $< $(FIX_BROKEN_CPP) -o $@
+	@$(CPP) -x c -P $< $(FIX_BROKEN_CPP) > $@
 
 %.ml: %.mlp $(BITSTRING)/pa_bitstring.cmo
 	$(CAMLP4OF) build/bitstring.cma $(BITSTRING)/bitstring_persistent.cmo $(BITSTRING)/pa_bitstring.cmo -impl $< -o $@
