--- ocamlnet-3.5.1.orig/src/netsys/netsys_posix.ml	2011-10-12 14:09:05.000000000 +0200
+++ ocamlnet-3.5.1/src/netsys/netsys_posix.ml	2012-01-12 19:33:39.000000000 +0100
@@ -412,9 +412,11 @@
 type at_flag = AT_EACCESS | AT_SYMLINK_NOFOLLOW | AT_REMOVEDIR
 
 (* The stubs assume these type definitions: *)
+(* In fact, they don't: they assume OCaml's stdlib definition
 type open_flag1 = Unix.open_flag =
     O_RDONLY | O_WRONLY | O_RDWR | O_NONBLOCK | O_APPEND | O_CREAT | O_TRUNC
   | O_EXCL | O_NOCTTY | O_DSYNC | O_SYNC | O_RSYNC
+*)
 
 type access_permission1 = Unix.access_permission =
     R_OK | W_OK | X_OK | F_OK
--- ocamlnet-3.5.1.orig/src/netstring/Makefile.def	2012-02-29 19:02:52.000000000 +0100
+++ ocamlnet-3.5.1/src/netstring/Makefile.def	2012-05-25 16:59:56.000000000 +0200
@@ -13,7 +13,7 @@
 PKGNAME  = netstring
 
 REQUIRES = $(REGEXP_PROVIDER) bigarray
-INCLUDES += $(INC_NETSYS)
+INCLUDES += $(INC_NETSYS) -I +compiler-libs
 
 ISO_MAPPINGS   = mappings/iso*.unimap
 JP_MAPPINGS    = mappings/jis*.*map
--- ocamlnet-3.5.1.orig/src/pop/netpop.ml	2012-02-29 19:02:53.000000000 +0100
+++ ocamlnet-3.5.1/src/pop/netpop.ml	2013-06-20 14:06:11.000000000 +0200
@@ -231,6 +231,7 @@
 	  status_response ic parse_line (Hashtbl.create 1)
     with _ -> raise Protocol_error
 
+(*
   method stat () =
     self#check_state `Transaction;
     send_command oc "STAT";
@@ -242,4 +243,5 @@
 	(count, size, ext)
       )
     with _ -> raise Protocol_error;
+*)
 end
--- ocamlnet-3.5.1/src/netstring/netencoding.mli.orig	2013-11-27 14:41:37.000000000 +0100
+++ ocamlnet-3.5.1/src/netstring/netencoding.mli	2013-11-27 14:41:52.000000000 +0100
@@ -120,7 +120,7 @@
 	 * to ensure that all output lines have a length <= 76 bytes.
 	 *
 	 * Note unsafe characters:
-	 *   As recommended by RFC 2045, the characters [!#$\@[]^`{|}~]
+	 *   As recommended by RFC 2045, the characters [!#$\@[]^`{}|~]
 	 *   and the double quotes
 	 *   are additionally represented as hex tokens.        
 	 *   Furthermore, the letter 'F' is considered as unsafe if it
